demo

```diff
@@ -62,16 +80,138 @@ if [[ "$?" -eq 1 ]]; then
     docker-compose pull ${compose_args}
 fi
 
-if [[ "$1" == "up" ]]; then
-    echo "Starting Mender demo environment.."
-fi
-
 # Pass this value on to the GUI container as an env variable
 export INTEGRATION_VERSION=$(git describe --tags --abbrev=0)
 
-exec docker-compose \
-     -f docker-compose.yml \
-     -f docker-compose.storage.minio.yml \
-     -f docker-compose.demo.yml \
-     $CLIENT \
-     "$@"
+if [[ "${!#}" != "up" ]]; then
+    # exec steals the shell, so unless docker-compose is not found,
+    # exit 1 will never happen.
+    exec docker-compose \
+         -f docker-compose.yml \
+         -f docker-compose.storage.minio.yml \
+         -f docker-compose.demo.yml \
+         $CLIENT \
+         "$@"
+    exit 1
+fi
+
+# ------------------------------------------------------------------------------
+#
+#       The following code will only be run in the case ./demo up [args]
+#
+# ------------------------------------------------------------------------------
+
+DOCKER_COMPOSE_PROJECT_NAME="${DOCKER_COMPOSE_PROJECT_NAME:-integration}"
+
+# Make sure that the demo environment is brought down on SIGINT
+exitfunc() {
+    retval=$(docker-compose \
+         -f docker-compose.yml \
+         -f docker-compose.storage.minio.yml \
+         -f docker-compose.demo.yml \
+         -p ${DOCKER_COMPOSE_PROJECT_NAME} \
+         $CLIENT \
+         down)
+    exit $retval
+}
+
+trap exitfunc SIGINT
+
+echo "Starting the Mender demo environment..."
+
+docker-compose \
+    -f docker-compose.yml \
+    -f docker-compose.storage.minio.yml \
+    -f docker-compose.demo.yml \
+    $CLIENT \
+    "$@" -d
+
+USER='mender-demo@example.com'
+PASSWORD=$(< /dev/urandom tr -dc A-Za-z0-9 | head -c12)
+
+MENDER_SERVER_USER="${USER}:${PASSWORD}"
+MENDER_SERVER_URI='https://localhost'
+
+# Block until the useradm service returns an HTTP response.
+# Which means that it is up.
+echo "Waiting for the useradm service to start up"
+curl --silent -k -X POST -u $MENDER_SERVER_USER\
+     --fail\
+     --retry-connrefused\
+     --retry 60\
+     --retry-delay 2\
+     --retry-max-time 120\
+     $MENDER_SERVER_URI/api/management/v1/useradm/auth/login > /dev/null
+
+retval=$?
+case $retval in
+    0)  ;; # User exists - continue.
+    22) ;; # Server 400 error, ie, server is up - continue.
+    *) echo "Failed to access the mender-useradm service"
+       echo "Please check the docker-compose setup."
+       echo "curl exit code" $retval
+       exit $retval
+       ;;
+esac
+
+echo "Creating a new user..."
+
+# Create a new user.
+docker exec ${DOCKER_COMPOSE_PROJECT_NAME}_mender-useradm_1 /usr/bin/useradm create-user --username=${USER} --password=${PASSWORD}
+
+retval=$?
+case $retval in
+     0) echo "New user created." ;;
+     5) echo "The user already exists"; exit 1;; # No need to continue if a demo user exists
```

Comment:
I don't understand. Do you assume that if the demo user exists the artifact has been already uploaded?

Reply:
Yes. That is the assumption. You want it explicitly checked?

Reply:
The reason it is done like this, is that a password is generated randomly for each user, so if the user exists already, we don't have the password. Maybe there should be an option for passing in a password through the command line?

Reply:
Now I see the problem, good point. I was just wondering about what if the user is created correctly but the Artifact is not uploaded for some reason. Then the user will need to run `./demo down` before retrying?

Maybe we can add such information in the error message: "run this if you want to delete existing user and then rerun this to create a new one".

Reply:
Yes, good idea. I'll incorporate the error message now.